ms_active_directory/core/ad_session.py

```diff
@@ -900,6 +979,25 @@ def find_groups_by_common_name(self, group_name: str, attributes_to_lookup: List
         logger.info('%s groups found with common name %s', len(res), group_name)
         return res
 
+    def find_groups_in_ou(self, ou_name: str, attributes_to_lookup: List[str] = None,
+                          controls: List[Control] = None) -> List[ADGroup]:
+        """ Find all groups within a given organizational unit and return a list of ADGroup objects.
+
+        :param ou_name: The name of the organizational unit where the group(s) are located.
+        :param attributes_to_lookup: A list of additional LDAP attributes to query for the group. Regardless of
+                                     what's specified, the groups' name and object class attributes will be queried.
+        :param controls: A list of LDAP controls to use when performing the search. These can be used to specify
+                         whether or not certain properties/attributes are critical, which influences whether a search
+                         may succeed or fail based on their availability.
+        :returns: a list of ADGroup objects representing groups with the specified organizational unit.
+        """
+        ou_dn = ldap_utils.construct_ou_distinguished_name(ou_name, self.domain_dns_name)
+        # a size limit of 0 means unlimited
+        res = self._find_ad_objects_and_attrs(ou_dn, ldap_constants.FIND_GROUP_FILTER, SUBTREE,
+                                              attributes_to_lookup, 0, ADGroup, controls)
+        logger.info('%s groups found within organizational unit %s', len(res), ou_name)
+        return res
```

Comment:
at present, doesn't really add new functionality. there's actually a `set_domain_search_base` that can be used if you want to confine your search to a specific OU.

this is helpful if you have multiple threads/processes all sharing a session object, but if we're doing that we might want to just add a `search_location` keyword argument to all of the existing `find_*` functions.

you can find all groups by doing a `find_groups_with_attribute` and using `objectClass` as the attribute with `*` as the value, so this is really a matter of just confining the search space for efficiency/security reasons, right?

Reply:
> this is helpful if you have multiple threads/processes all sharing a session object, but if we're doing that we might want to just add a `search_location` keyword argument to all of the existing find_* functions.

I think this would be best and would account for a broader use case. Our particular use case is we have several types of groups that are differentiated by being placed in various top level OU's. A `search_location` argument would definitely account for that.

Reply:
I can go through and add it to the various `find_` functions this weekend 😄